doseCalc/+DoseEngines/matRad_ParticlePencilBeamEngineAbstract.m

```diff
@@ -0,0 +1,838 @@
+            %use indexing for performance and avoid to many copies
+            bixel = this.getBixelIndicesOnRay(bixel,currRay);
+
+            if isempty(bixel.ix)
+                return;
+            end     
+
+            % Get quantities 1:1 from ray. Here we trust Matlab's memory
+            % management to not copy the arrays until they are modified.
+            % This allows us to efficiently access them by indexing in the
+            % bixel computation
+            bixel.radialDist_sq = currRay.radialDist_sq(bixel.subRayIx);
+            bixel.radDepths = currRay.radDepths(bixel.subRayIx);
+            if this.calcBioDose
+                bixel.vTissueIndex = currRay.vTissueIndex(bixel.subRayIx);
+            end
+        
+        end
+
+        function X = interpolateKernelsInDepth(this,bixel)
+            baseData = bixel.baseData;
+            
+            depths = baseData.depths;
+
+            % add potential offset
+            if isfield(baseData,'offset')
+                depths = depths + baseData.offset - bixel.radDepthOffset;
+            end
+            
+            % calculate particle dose for bixel k on ray j of beam i
+            % convert from MeV cm^2/g per primary to Gy mm^2 per 1e6 primaries
+            conversionFactor = 1.6021766208e-02;
+            
+            %Find all values we need to interpolate
+            X.Z = conversionFactor*baseData.Z;
+
+             %Single or double gaussian
+            if ~isfield(baseData,'sigma')
+                X.sigma1 = baseData.sigma1;
+                X.sigma2 = baseData.sigma2;
+                X.weight = baseData.weight;
+            else
+                X.sigma  = baseData.sigma;
+            end
+
+            % LET
+            if this.calcLET
+                X.LET = baseData.LET;
+            end
+
+            % bioDose
+            if this.calcBioDose
+                X.alpha = baseData.alpha;
+                X.beta = baseData.beta;
+            end   
+            
+            X = structfun(@(v) matRad_interp1(depths,v,bixel.radDepths,'nearest'),X,'UniformOutput',false); %Extrapolate to zero?           
+        end
+
+        % We override this function to boost efficiency a bit (latDistX & Z
+        % not needed)
+        function ray = getRayGeometryFromBeam(this,ray,currBeam)
+            lateralRayCutOff = this.getLateralDistanceFromDoseCutOffOnRay(ray);
+
+            % Ray tracing for beam i and ray j
+            [ix,ray.radialDist_sq] = this.calcGeoDists(currBeam.bevCoords, ...
+                ray.sourcePoint_bev, ...
+                ray.targetPoint_bev, ...
+                ray.SAD, ...
+                currBeam.ixRadDepths, ...
+                lateralRayCutOff);
+            
+            ray.geoDepths = currBeam.geoDepths{1}(ix);
+            ray.radDepths = currBeam.radDepths{1}(ix);
+            ray.ix = currBeam.ixRadDepths(ix);
+            ray.subIxVdoseGrid = currBeam.subIxVdoseGrid(ix);
```

Comment:
this line is the only additional line wrt superclass function `matRad_PencilBeamEngineAbstract.getRayGeometryFromBeam()`. Other lines are identical, could simply call the superclass function with `getRayGeometryFromBeam@superclassName`

Reply:
I think it is not so easy. The problem are the return options of the calcGeoDists function. Our Particle pencil-beam algorithm does not need distances in x/y, but only the radial distance. Finesampling needs x/y lateral distances, and photons need the lateral distances projected onto the isocenter plane. So each of these functions uses different calls ot the geodists. but maybe there is a better way to modularize, I will investigate.